src/Storages/S3Queue/S3QueueSource.cpp

```diff
@@ -0,0 +1,339 @@
+#include <algorithm>
+#include <Common/ProfileEvents.h>
+#include <Common/ZooKeeper/ZooKeeper.h>
+#include "IO/ParallelReadBuffer.h"
+#include "Parsers/ASTCreateQuery.h"
+#include "config.h"
+
+#if USE_AWS_S3
+
+#    include <Common/isValidUTF8.h>
+
+#    include <Functions/FunctionsConversion.h>
+
+#    include <IO/S3/Requests.h>
+#    include <IO/S3Common.h>
+
+#    include <Interpreters/TreeRewriter.h>
+
+#    include <Parsers/ASTFunction.h>
+#    include <Parsers/ASTInsertQuery.h>
+
+#    include <Storages/NamedCollectionsHelpers.h>
+#    include <Storages/PartitionedSink.h>
+#    include <Storages/ReadFromStorageProgress.h>
+#    include <Storages/S3Queue/S3QueueSource.h>
+#    include <Storages/StorageS3.h>
+#    include <Storages/StorageS3Settings.h>
+#    include <Storages/VirtualColumnUtils.h>
+#    include <Storages/getVirtualsForStorage.h>
+
+#    include <Formats/FormatFactory.h>
+
+#    include <Processors/Formats/IInputFormat.h>
+#    include <Processors/Formats/IOutputFormat.h>
+#    include <Processors/Transforms/AddingDefaultsTransform.h>
+
+#    include <QueryPipeline/QueryPipelineBuilder.h>
+
+#    include <DataTypes/DataTypeString.h>
+
+#    include <Common/CurrentMetrics.h>
+#    include <Common/NamedCollections/NamedCollections.h>
+#    include <Common/parseGlobs.h>
+
+#    include <Processors/ISource.h>
+#    include <Processors/Sinks/SinkToStorage.h>
+
+
+namespace CurrentMetrics
+{
+extern const Metric StorageS3Threads;
+extern const Metric StorageS3ThreadsActive;
+}
+
+namespace ProfileEvents
+{
+extern const Event S3DeleteObjects;
+extern const Event S3ListObjects;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int S3_ERROR;
+}
+
+
+StorageS3QueueSource::QueueGlobIterator::QueueGlobIterator(
+    const S3::Client & client_,
+    const S3::URI & globbed_uri_,
+    ASTPtr query,
+    const Block & virtual_header,
+    ContextPtr context,
+    UInt64 & max_poll_size_,
+    StorageS3QueueSource::KeysWithInfo * read_keys_,
+    const S3Settings::RequestSettings & request_settings_)
+    : max_poll_size(max_poll_size_)
+    , bucket(globbed_uri_.bucket)
+    , glob_iterator(std::make_unique<StorageS3QueueSource::DisclosedGlobIterator>(
+          client_, globbed_uri_, query, virtual_header, context, read_keys_, request_settings_))
+{
+    while (true)
+    {
+        KeyWithInfo val = glob_iterator->next();
+        if (val.key.empty())
+        {
+            break;
+        }
+        keys_buf.push_back(val);
+    }
```

Comment:
Looks a bit scary that this loop can be really long in case we have a log of files in s3.
What does it mean that `val.key.empty()`?
Can we avoid making this cycle here and move it to `next` method?